### PR TITLE
[WIP] Python Fundamentals

### PR DESCRIPTION
**Pull request by mlp6**
* Lecture outline available
* Skeleton for next assignment ready for some eyes on it (ipynb is missing).
* Will put together ipynb in morning; @suyashkumar need suggestion on a way to easily evaluate completion.

### CODE REVIEW

File: Assignments/03_python_fundamentals.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+## Mini-project
+* We are going to start putting your `git` and `python` skills to use!
+* The repository that you forked last lecture has hall all of the Pull Requests
+  from each student merged into the master branch.
+* Update the `master` branch of your forked repository to reflect the commits
+  up until the `v1.0.0` annotated tag (which should be the latest version).
+  Note that you can do this through GitHub **or** locally in your forked
+  repository from the CLI.  Try doing it through the CLI by creating a second
+  remote called `upstream` that use the URL of the `mlp6` version of the
```
Comment (suyashkumar) [inline Assignments/03_python_fundamentals.md]: second remotes may be a little confusing for students right now...but would be instructive for them to pull through with this one.

File: Assignments/03_python_fundamentals.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  section of the Python Fundamentals lecture, write a Python program that (on a
+  git feature branch!):
+  + Concatenates all of the individual CSV files--execpt for `mlp6.csv`--into a
+    single file called `everyone.csv`.
+  + Checks that there are no spaces in the team name that a student specified.
+  + Counts how many team names used CamelCase, and prints that could to
+    `STDOUT` when the program finishes writing the file.
+  + In addition to writing a CSV file, create a JSON-formatted file called
+    `everyone.json`.
+* Submit a Pull Request that has your modular code, along with `everyone.csv`
```
Comment (suyashkumar) [inline Assignments/03_python_fundamentals.md]: +1

File: Lectures/PythonFundamentals.md [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+* Python has a stadard library of functionality; additional functionality is
+  imported into projects through additional packages.
+  + PyPI is the most popular repository of Python packages.
+  + `pip` is the tool to install packages from PyPI.
+  + Different projects will require different packages.  Installing just the packages that you need for each Python project:
+    - Prevents package version conflicts between different projects.
+    - Establishes a minimized storage footprint of packages for each project
+      (though commonly-used packages in many projects will be replicated).
+  + `virtualenv` will be used to create virtual environments for each Python project.
+  + :eyes: Make sure that you have `pip` and `virtualenv` installed on your laptop. :eyes:
```
Comment (suyashkumar) [inline Lectures/PythonFundamentals.md]: not sure if `virtualenv` comes with the default python installation. Folks may have to install it

File: Assignments/03_python_fundamentals.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+## Exploring Python Syntax
+* Clone this repository (or fork is you might want to submit a Pull Request with changes at some point): https://github.com/mlp6/Medical-Software-Design
+* Create a local virtual environment with the packages in `requirements.txt`.
+* Activate your virtual environment.
+* Change into the `Assignments` directory and run `jupyter notebook
+  PythonFundamentals.ipynb`.  You will be prompted to launch a web browser that
+  has an address that starts with `127.0.0.1` or `localhost`; your web browser
+  might automatically be launched with this page too.
+* Work through the interactive notebook exercise.
+* **TODO: Come up with a way to evaluate this is completed.**
```
Comment (suyashkumar) [inline Assignments/03_python_fundamentals.md]: We can have students upload a completed ipynb file to a git repo for us to inspect (they can modify some things, and maybe make some computations based on the time they run it. If we want to really check for completeness, one of the cells to be run can compute a hash of the MAC address for the machine the notebook was run on).

Comment (mlp6) [reply Assignments/03_python_fundamentals.md]: Decided to have the render a PDF of the completed notebook, just as an honesty check.

Comment (suyashkumar) [reply Assignments/03_python_fundamentals.md]: Cool, if you wanted to discourage folks from turning in the same PDF you could include a cell that calls and prints the following:
```
from uuid import getnode
getnode()
```
This returns the MAC address as a 48 bit integer.

**Pull request merged**